resolver/resolve.go

```diff
@@ -392,21 +392,32 @@ resolveLoop:
 	}
 
 	// Check if we want to use an older cache instead.
-	switch {
-	case err != nil:
-		// There was an error during resolving, return the old cache entry instead.
-		return oldCache, nil
-	case rrCache.IsNXDomain():
-		// The new result is NXDomain, return the old cache entry instead.
-		return oldCache, nil
+	if oldCache != nil {
+		oldCache.isBackup = true
+
+		switch {
+		case err != nil:
+			// There was an error during resolving, return the old cache entry instead.
+			log.Tracer(ctx).Debugf("resolver: serving backup cache of %s because query failed: %s", q.ID(), err)
+			return oldCache, nil
+		case rrCache.IsNXDomain():
+			// The new result is NXDomain, return the old cache entry instead.
+			log.Tracer(ctx).Debugf("resolver: serving backup cache of %s because fresh response is NXDomain", q.ID())
+			return oldCache, nil
+		}
```

Comment:
I know this shouldn't harm much but we should returns NXDOMAIN to the user instead of serving a stale record as the IP address is likely not correct anymore. We should serve the cache for SERVFAIL but IMHO not for NXDOMAIN.